File: slr/R/fastar1.R
# Copyright 2010 Nathan M. Urban

source("roblib.R")
dynReload("fastar1")

# generate AR(1) time series of length n with lag-1 autocorrelation coefficient rho and innovation variance sigma^2
ar1.sim = function(n, rho, sigma)
{
    w = rnorm(n)
    
    cout = .C( "ar1sim",
                n = as.integer(n),
                rho = as.double(rho),
                sigma = as.double(sigma),
                w = as.double(w),
                r = as.double(rep(0,n))
            )
    
    return(cout$r)
}

File: lib/R/fastar1.c
/*   Copyright 2010 Nathan M. Urban                                           */
/*   build with:  system("rm fastar1.o; R CMD SHLIB fastar1.c")               */

/*   Simulates AR(1) red noise                                                */

/*   n = length of time series to generate                                    */
/*   rhoin = (pointer to) lag-1 autocorrelation coefficient                   */
/*   sigmain = (pointer to) innovation standard deviation                     */
/*   w = time series array of iid standard normal white noise (of length n)   */
/*   r = time series array of AR1 red noise (of length n)                     */

#include <math.h>

void ar1sim(int* nin, double* rhoin, double* sigmain, double* w, double* r)
{
  double rho, sigma, sigma_proc;
  int n, i;
  
  n = nin[0]; rho = rhoin[0]; sigma = sigmain[0];

  sigma_proc = sigma/sqrt(1.0-rho*rho); /* process variance */
  r[0] = sigma_proc*w[0]; /* draw initial value from stationary distribution */

  for(i=1; i<n; i++)
    r[i] = rho*r[i-1] + sigma*w[i];
}
